File: frontend/src/types/client-portal.ts
export type ClientProjectOverview = {
  project_id: string;
  project_name: string;
  description: string | null;
  current_status: string;
  overall_health: "green" | "amber" | "red" | "insufficient";
  delivery_confidence: number | null;
  delivery_confidence_label: string;
  current_phase: string;
  completion_percentage: number;
  start_date: string;
  target_end_date: string;
};

export type ClientMilestone = {
  id: string;
  name: string;
  description: string | null;
  planned_date: string;
  actual_date: string | null;
  status: string;
  progress_percentage: number;
};

export type ClientRisk = {
  id: string;
  title: string;
  severity: string;
  impact: string | null;
  mitigation: string | null;
  status: string;
  updated_at: string;
};

export type ClientAction = {
  id: string;
  title: string;
  description: string | null;
  action_type: "approval" | "information_request" | "client_action";
  due_date: string | null;
  status: string;
  is_overdue: boolean;
};

export type ClientPortalReport = {
  id: string;
  title: string;
  report_type: string;
  executive_summary: string;
  published_at: string;
  pdf_download_url: string;
  csv_download_url: string;
};

export type ClientAiSummary = {
  title: string;
  summary: string;
  current_progress: string[];
  risks: string[];
  upcoming_work: string[];
  generated_at: string | null;
};

export type ClientDocument = {
  id: string;
  title: string;
  document_type: string;
  version: string;
  description: string | null;
  file_name: string;
  file_url: string | null;
  shared_at: string;
};

export type ClientDeliverable = {
  id: string;
  title: string;
  description: string | null;
  status: "completed" | "in_progress" | "planned";
  due_date: string | null;
  completed_at: string | null;
  file_name: string | null;
  file_url: string | null;
};

export type ClientChangeRequestStatus =
  | "submitted"
  | "under_review"
  | "approved"
  | "rejected"
  | "implemented";

export type ClientChangeRequest = {
  id: string;
  project_id: string;
  title: string;
  description: string;
  business_justification: string | null;
  priority: "low" | "medium" | "high" | "critical";
  status: ClientChangeRequestStatus;
  decision_notes: string | null;
  implemented_at: string | null;
  created_at: string;
  updated_at: string;
};

export type ClientChangeRequestCreate = {
  title: string;
  description: string;
  business_justification?: string | null;
  priority: "low" | "medium" | "high" | "critical";
};

export type ClientMeetingActionItem = {
  title: string;
  owner: string | null;
  due_date: string | null;
  status: string;
};

export type ClientMeeting = {
  id: string;
  title: string;
  starts_at: string;
  duration_minutes: number;
  meeting_url: string | null;
  agenda: string | null;
  minutes: string | null;
  action_items: ClientMeetingActionItem[];
  status: "scheduled" | "completed" | "cancelled";
};

export type ClientNotification = {
  id: string;
  notification_type:
    | "report_published"
    | "milestone_completed"
    | "risk_updated"
    | "document_shared"
    | "meeting_scheduled";
  title: string;
  detail: string | null;
  occurred_at: string;
  href: string | null;
};

export type ClientProjectDashboard = {
  overview: ClientProjectOverview;
  milestones: ClientMilestone[];
  risks: ClientRisk[];
  client_actions: ClientAction[];
  reports: ClientPortalReport[];
  ai_summary: ClientAiSummary;
  documents: ClientDocument[];
  deliverables: ClientDeliverable[];
  change_requests: ClientChangeRequest[];
  meetings: ClientMeeting[];
  notifications: ClientNotification[];
  generated_at: string;
};

File: backend/app/schemas/client_portal.py
from datetime import date, datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field

from app.schemas.common import ORMModel


class ClientProjectOverviewRead(BaseModel):
    project_id: UUID
    project_name: str
    description: str | None = None
    current_status: str
    overall_health: Literal["green", "amber", "red", "insufficient"]
    delivery_confidence: int | None = None
    delivery_confidence_label: str
    current_phase: str
    completion_percentage: int = Field(ge=0, le=100)
    start_date: date
    target_end_date: date


class ClientMilestoneRead(BaseModel):
    id: UUID
    name: str
    description: str | None = None
    planned_date: date
    actual_date: date | None = None
    status: str
    progress_percentage: int = Field(ge=0, le=100)


class ClientRiskRead(BaseModel):
    id: UUID
    title: str
    severity: str
    impact: str | None = None
    mitigation: str | None = None
    status: str
    updated_at: datetime


class ClientActionRead(BaseModel):
    id: UUID
    title: str
    description: str | None = None
    action_type: Literal["approval", "information_request", "client_action"]
    due_date: date | None = None
    status: str
    is_overdue: bool = False


class ClientReportRead(BaseModel):
    id: UUID
    title: str
    report_type: str
    executive_summary: str
    published_at: datetime
    pdf_download_url: str
    csv_download_url: str


class ClientAiSummaryRead(BaseModel):
    title: str
    summary: str
    current_progress: list[str] = Field(default_factory=list)
    risks: list[str] = Field(default_factory=list)
    upcoming_work: list[str] = Field(default_factory=list)
    generated_at: datetime | None = None


class ClientDocumentRead(BaseModel):
    id: UUID
    title: str
    document_type: str
    version: str
    description: str | None = None
    file_name: str
    file_url: str | None = None
    shared_at: datetime


class ClientDeliverableRead(BaseModel):
    id: UUID
    title: str
    description: str | None = None
    status: str
    due_date: date | None = None
    completed_at: datetime | None = None
    file_name: str | None = None
    file_url: str | None = None


class ClientChangeRequestCreate(BaseModel):
    title: str = Field(min_length=3, max_length=160)
    description: str = Field(min_length=10, max_length=5000)
    business_justification: str | None = Field(default=None, max_length=3000)
    priority: Literal["low", "medium", "high", "critical"] = "medium"


class ClientChangeRequestRead(ORMModel):
    id: UUID
    project_id: UUID
    title: str
    description: str
    business_justification: str | None = None
    priority: str
    status: str
    decision_notes: str | None = None
    implemented_at: datetime | None = None
    created_at: datetime
    updated_at: datetime


class ClientMeetingActionItemRead(BaseModel):
    title: str
    owner: str | None = None
    due_date: date | None = None
    status: str = "open"


class ClientMeetingRead(BaseModel):
    id: UUID
    title: str
    starts_at: datetime
    duration_minutes: int
    meeting_url: str | None = None
    agenda: str | None = None
    minutes: str | None = None
    action_items: list[ClientMeetingActionItemRead] = Field(default_factory=list)
    status: str


class ClientNotificationRead(BaseModel):
    id: str
    notification_type: Literal[
        "report_published",
        "milestone_completed",
        "risk_updated",
        "document_shared",
        "meeting_scheduled",
    ]
    title: str
    detail: str | None = None
    occurred_at: datetime
    href: str | None = None


class ClientProjectDashboardRead(BaseModel):
    overview: ClientProjectOverviewRead
    milestones: list[ClientMilestoneRead]
    risks: list[ClientRiskRead]
    client_actions: list[ClientActionRead]
    reports: list[ClientReportRead]
    ai_summary: ClientAiSummaryRead
    documents: list[ClientDocumentRead]
    deliverables: list[ClientDeliverableRead]
    change_requests: list[ClientChangeRequestRead]
    meetings: list[ClientMeetingRead]
    notifications: list[ClientNotificationRead]
    generated_at: datetime
